server/data/data-jmap/src/test/java/org/apache/james/jmap/api/preview/MessagePreviewStoreContract.java

```diff
@@ -120,37 +120,40 @@ default void storeShouldBeIdempotent() {
     }
 
     @Test
-    default void concurrentStoreShouldStoreTheLatestValueDespiteOfCollisions() throws Exception {
+    default void concurrentStoreShouldOverrideOldValueWhenSameMessageId() throws Exception {
         int threadCount = 10;
         int stepCount = 100;
-        int repeatCount = 10;
-
-        ConcurrentHashMap<Integer, Integer> indexCounter = new ConcurrentHashMap<>();
-
-        Flux.range(0, stepCount)
-            .repeat(repeatCount)
-            .parallel(threadCount)
-            .runOn(Schedulers.newParallel("preview", threadCount))
-            .flatMap(index -> {
-                Integer counter = indexCounter.compute(index, (id, count) -> {
-                    if (count == null) {
-                        return 0;
-                    }
-                    return count + 1;
-                });
-
-                return Mono.from(testee()
-                    .store(TestMessageId.of(index), Preview.from(String.valueOf(counter))))
-                    .then();
-            })
-            .then()
-            .block();
-
-        IntStream.range(0, stepCount)
+
+        ConcurrentTestRunner.builder()
+            .operation((thread, step) -> Mono.from(testee()
+                .store(TestMessageId.of(thread), Preview.from(String.valueOf(step))))
+                .block())
+            .threadCount(threadCount)
+            .operationCount(stepCount)
+            .runSuccessfullyWithin(Duration.ofMinutes(1));
+
+        IntStream.range(0, threadCount)
             .forEach(index -> assertThat(Mono.from(testee()
                     .retrieve(TestMessageId.of(index)))
                     .block())
-                .isEqualTo(Preview.from(String.valueOf(repeatCount))));
+                .isEqualTo(Preview.from(String.valueOf(stepCount - 1))));
+    }
+
+    @Test
+    default void concurrentStoreOnSameMessageIdShouldStoreTheLatestValue() throws Exception {
+        AtomicInteger previewValue = new AtomicInteger();
+
+        ConcurrentTestRunner.builder()
+            .operation((thread, step) -> Mono.from(testee()
+                .store(MESSAGE_ID_1, Preview.from(String.valueOf(previewValue.incrementAndGet()))))
```

Comment:
AGain to atomic operation don't combine to a single atomic big operation (even if they are inlined)